File: blaschke3.m
%Victor del Valle del Apio
%victorvalleapio@gmail.com
%Julio de 2015

%Aplicacion de los productos de Blaschke de grado tres

function varargout = guiBlaschke3(varargin)
    gui_Singleton = 1;
    gui_State = struct('gui_Name',       mfilename, ...
                       'gui_Singleton',  gui_Singleton, ...
                       'gui_OpeningFcn', @guiBlaschke3_OpeningFcn, ...
                       'gui_OutputFcn',  @guiBlaschke3_OutputFcn, ...
                       'gui_LayoutFcn',  [] , ...
                       'gui_Callback',   []);
    if nargin && ischar(varargin{1})
        gui_State.gui_Callback = str2func(varargin{1});
    end

    if nargout
        [varargout{1:nargout}] = gui_mainfcn(gui_State, varargin{:});
    else
        gui_mainfcn(gui_State, varargin{:});
    end
end

function guiBlaschke3_OpeningFcn(hObject, eventdata, handles, varargin)
    handles.output = hObject;
    guidata(hObject, handles);
    pushbutton1_Callback(hObject, eventdata, handles);
end

function varargout = guiBlaschke3_OutputFcn(hObject, eventdata, handles) 
    varargout{1} = handles.output;
end

%Funcion principal
function pushbutton1_Callback(hObject, eventdata, handles)
    %Limpiamos los ejes
    cla
    hold on
    axis equal
    
    %Obtenemos los valores de a1 y a2 introducidos
    a1 = str2double(get(handles.edit1, 'String'));
    a2 = str2double(get(handles.edit2, 'String'));
    
    %Si a1 y a2 se encuentran dentro del disco unidad:
    if (norm(a1)<1 && norm(a2)<1)  
        set(handles.text3, 'String', ['Producto de Blaschke donde a1 = ' num2str(a1) ' y a2 = ' num2str(a2)])
        
        %Dibujamos a1 y a2
        plot(real(a1), imag(a1), '-*b')
        plot(real(a2), imag(a2), '-*b')
        
        %Para cada punto del disco unidad, calculamos sus tres imagenes inversas
        for lambda=0:0.5:2*pi
            z = exp(1i*lambda);
            
            %Resolvemos la ecuacion para calcular las tres raices
            p = [1 -(a1+a2+conj(a1)*conj(a2)*z) (a1*a2+conj(a1)*z+conj(a2)*z) -z];
            r = roots(p);
            
            %Con las tres raices, dibujamos el triangulo que forman
            for ind = 1:length(r)
                m(ind, 1) = real(r(ind));
                m(ind, 2) = imag(r(ind));
            end
            plot([m(1,1) m(2,1)],[m(1,2) m(2,2)],'r')
            plot([m(1,1) m(3,1)],[m(1,2) m(3,2)],'r')
            plot([m(2,1) m(3,1)],[m(2,2) m(3,2)],'r')
        end

        %Dibujamos el disco unidad
        theta_grid = linspace(0,2*pi);
        circ_x = cos(theta_grid);
        circ_y = sin(theta_grid);
        plot(circ_x, circ_y,'b')

        %Si alguno de los puntos no estaba en el disco unidad:
        elseif(norm(a1)>=1 && norm(a2)<1)
            set(handles.text3, 'String', ['El punto a1 = ' num2str(a1) ' no está dentro del disco unidad'])
        elseif(norm(a1)<1 && norm(a2)>=1)
            set(handles.text3, 'String', ['El punto a2 = ' num2str(a2) ' no está dentro del disco unidad'])
        else
            set(handles.text3, 'String', ['Los puntos a1 = ' num2str(a1) ' y a2 = ' num2str(a2) ' no están dentro del disco unidad'])
    end

    hold off
end

%Input de a1
function edit1_Callback(hObject, eventdata, handles)
    pushbutton1_Callback(hObject, eventdata, handles)
end

function edit1_CreateFcn(hObject, eventdata, handles)
    if ispc && isequal(get(hObject,'BackgroundColor'), get(0,'defaultUicontrolBackgroundColor'))
        set(hObject,'BackgroundColor','white');
    end
end

%Input de a2
function edit2_Callback(hObject, eventdata, handles)
    pushbutton1_Callback(hObject, eventdata, handles)
end

function edit2_CreateFcn(hObject, eventdata, handles)
    if ispc && isequal(get(hObject,'BackgroundColor'), get(0,'defaultUicontrolBackgroundColor'))
        set(hObject,'BackgroundColor','white');
    end
end

File: blaschke3.py
#Victor del Valle del Apio
#victorvalleapio@gmail.com
#Julio de 2015

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Polygon,Circle

#Clase principal del programa
class Blaschke3 :
    
    #Metodo para inicializar las variables 
    def __init__(self,ax,fig) :
        self.polygon = []
        self.cont = 0
        self.polygon_exist = False
        self.num_points = 20
        self.touched_circle = False
        self.list_circles = []
        self.ax = ax
        self.fig = fig
        self.cid_press = fig.canvas.mpl_connect('button_press_event', self.on_press)
        self.cid_move = fig.canvas.mpl_connect('motion_notify_event', self.on_move)
        self.cid_release = fig.canvas.mpl_connect('button_release_event', self.on_release)

    #Metodo que se activa cuando se pulsa un punto
    def on_press(self, event):
        #Comprobamos si hemos pulsado en un punto antiguo para moverlo
        for circle in self.list_circles:
            contains, attr = circle.contains(event)
            if contains:
                self.touched_circle = circle
                self.exists_touched_circle = True
                self.pressed_event = event
                self.touched_x0, self.touched_y0 = circle.center
                return
                
        #Limpiamos los ejes y pintamos todos los puntos  
        plt.cla()        
        if self.cont==2:
            self.list_circles.remove(self.list_circles[0])
        c = Circle((event.xdata,event.ydata),0.02)
        ax.add_patch(Circle((0,0),1,color='blue', fill = False))
        self.list_circles.append(c)
        for circle in self.list_circles :
            self.ax.add_patch(circle)
        
        #Guardamos el nuevo punto (borrando el primero si ya habia dos)
        new_point = [event.xdata,event.ydata]
        if not(self.polygon_exist) :
            self.polygon.append(new_point)
            self.P = Polygon(self.polygon, closed = False,color = 'red', fill = False)
            self.polygon_exist = True
        else : 
            if self.cont==2:
                self.polygon.remove(self.polygon[0])
            self.polygon.append(new_point)
        self.P.set_xy(self.polygon)
        self.ax.add_patch(self.P)
        self.fig.canvas.draw()
        
        #Si a1 y a2 se encuentran dentro del disco unidad:
        if self.cont<2:
            self.cont=self.cont+1
        if self.cont==2:
            a1 = self.polygon[0]
            a2 = self.polygon[1]
            if np.linalg.norm(a1)<=1 and np.linalg.norm(a2)<=1:
                self.calculate_Blaschke3(a1,a2)

    #Metodo que se activa cuando se mueve un punto
    def on_move(self,event) : 
        
        if self.touched_circle and self.cont==2:
            dx = event.xdata - self.pressed_event.xdata
            dy = event.ydata - self.pressed_event.ydata
            x0, y0 = self.touched_circle.center
            self.touched_circle.center =  self.touched_x0 + dx, self.touched_y0 + dy
            plt.cla()
            ax.add_patch(Circle((0,0),1,color='blue', fill = False))
            for circle in self.list_circles :
                self.ax.add_patch(circle)
            self.polygon = [circle.center for circle in self.list_circles]
            self.P.set_xy(self.polygon)
            self.ax.add_patch(self.P)
            self.fig.canvas.draw()
            
            if self.cont==2:
                a1 = self.polygon[0]
                a2 = self.polygon[1]
                if np.linalg.norm(a1)<=1 and np.linalg.norm(a2)<=1:
                    self.calculate_Blaschke3(a1,a2)
        
    #Metodo que se activa cuando se suelta un punto
    def on_release(self, event) :
        self.touched_circle = False
    
    #Funcion principal
    def calculate_Blaschke3(self, a1, a2):
        a1 = complex(a1[0],a1[1])
        a2 = complex(a2[0],a2[1])
        t = np.linspace(0,2*np.pi,self.num_points)
        #Para cada punto del disco unidad, calculamos sus tres imagenes inversas        
        for interv in t:
            z = np.exp(interv*1j)
            
            #Resolvemos la ecuacion para calcular las tres raices
            a = 1
            b = -(a1+a2+a1.conjugate()*a2.conjugate()*z)
            c = (a1*a2+a1.conjugate()*z+a2.conjugate()*z)
            d = -z   
            solucion = np.roots([a, b, c, d])
            sol1=complex(solucion[0])
            sol2=complex(solucion[1])
            sol3=complex(solucion[2])
            
            #Dibujamos las 3 rectas
            plane=[]
            plane.append([sol1.real,sol1.imag])
            plane.append([sol2.real,sol2.imag])
            plane.append([sol3.real,sol3.imag])
            Pdraw = Polygon(plane, closed = True,color = 'red', fill = False)
            ax.add_patch(Pdraw)
           
#Funcion Main            
if __name__ == "__main__" :
    fig = plt.figure()
    ax = fig.add_subplot(111,aspect = 1)
    ax.set_xlim(-1,1)
    ax.set_ylim(-1,1)
    ax.add_patch(Circle((0,0),1,color='blue', fill = False))
    b3 = Blaschke3(ax,fig) 
    plt.show()
